internal/controllers/machinedeployment/machinedeployment_controller.go

```diff
@@ -286,8 +293,36 @@ func (r *Reconciler) reconcile(ctx context.Context, cluster *clusterv1.Cluster,
 	return errors.Errorf("unexpected deployment strategy type: %s", md.Spec.Strategy.Type)
 }
 
-// getMachineSetsForDeployment returns a list of MachineSets associated with a MachineDeployment.
-func (r *Reconciler) getMachineSetsForDeployment(ctx context.Context, md *clusterv1.MachineDeployment) ([]*clusterv1.MachineSet, error) {
+func (r *Reconciler) reconcileDelete(ctx context.Context, md *clusterv1.MachineDeployment) error {
+	log := ctrl.LoggerFrom(ctx)
+	msList, err := r.getAndAdoptMachineSetsForDeployment(ctx, md)
+	if err != nil {
+		return err
+	}
+
+	// If all the descendant machinesets are deleted, then remove the machinedeployment's finalizer.
+	if len(msList) == 0 {
+		controllerutil.RemoveFinalizer(md, clusterv1.MachineDeploymentFinalizer)
+		return nil
+	}
+
+	log.Info("MachineDeployment still has descendant MachineSets - deleting them first", "count", len(msList), "descendants", descendantMachineSets(msList))
```

Comment:
What about:
```suggestion
	log.Info("Waiting for MachineSets to be deletes", "MachineSets", descendantMachineSets(msList))
```
(dropping count, see my comment for descendantMachineSets)